src/legacy/core_plugins/kbn_vislib_vis_types/public/components/options/metrics_axes/index.tsx

```diff
@@ -0,0 +1,318 @@
+      setIsCategoryAxisHorizontal(isAxisHorizontal(chartPosition));
+
+      stateParams.valueAxes.forEach((axis, index) => {
+        if (isAxisHorizontal(axis.position) === isChartHorizontal) {
+          const position = mapPosition(axis.position);
+          onValueAxisPositionChanged(index, position);
+        }
+      });
+    },
+    [stateParams.valueAxes, onValueAxisPositionChanged]
+  );
+
+  const addValueAxis = useCallback(() => {
+    const lastAxisIdNumber = axesNumbers[VALUE_AXIS_PREFIX];
+    let nextAxisIdNumber;
+
+    if (!lastAxisIdNumber) {
+      nextAxisIdNumber = stateParams.valueAxes.reduce(countNextAxisNumber(VALUE_AXIS_PREFIX), 1);
+    } else {
+      nextAxisIdNumber = lastAxisIdNumber + 1;
+    }
+
+    const newAxis = cloneDeep(stateParams.valueAxes[0]);
+    newAxis.id = VALUE_AXIS_PREFIX + nextAxisIdNumber;
+    newAxis.position = mapPositionOpposite(newAxis.position);
+
+    const [name, lastAxisNameNumber] = getUpdatedAxisName(
+      newAxis.position,
+      axesNumbers,
+      stateParams.valueAxes
+    );
+    newAxis.name = name;
+
+    setAxesNumbers({
+      ...axesNumbers,
+      [VALUE_AXIS_PREFIX]: nextAxisIdNumber,
+      [newAxis.position]: lastAxisNameNumber,
+    });
+
+    setValue('valueAxes', [...stateParams.valueAxes, newAxis]);
+    return newAxis;
+  }, [stateParams.valueAxes, axesNumbers, getUpdatedAxisName, setValue]);
+
+  const removeValueAxis = useCallback(
+    (axis: ValueAxis) => {
+      const newValueAxes = stateParams.valueAxes.filter(valAxis => valAxis.id !== axis.id);
+
+      setValue('valueAxes', newValueAxes);
+
+      const updatedNumbers = getPreviousAxisNumbers(axis, axesNumbers);
+      if (updatedNumbers) {
+        setAxesNumbers({
+          ...updatedNumbers,
+        });
+      }
+
+      let chartIndex;
+      stateParams.seriesParams.forEach(({ valueAxis }, index) => {
+        if (axis.id === valueAxis) {
+          chartIndex = index;
+        }
+      });
+
+      if (chartIndex !== undefined) {
+        // if a seriesParam has valueAxis equals to removed one, then we reset it to the first valueAxis
+        setParamByIndex('seriesParams', chartIndex, 'valueAxis', newValueAxes[0].id);
+      }
+    },
+    [stateParams.seriesParams, stateParams.valueAxes, setParamByIndex, setValue]
+  );
+
+  const changeValueAxis: ChangeValueAxis = useCallback(
+    (index, paramName, selectedValueAxis) => {
+      let newValueAxis = selectedValueAxis;
+      if (selectedValueAxis === 'new') {
+        const axis = addValueAxis();
+        newValueAxis = axis.id;
+      }
+
+      setParamByIndex('seriesParams', index, paramName, newValueAxis);
+
+      updateAxisTitle();
+    },
+    [addValueAxis, setParamByIndex]
+  );
+
+  const metrics = useMemo(() => {
+    const schemaName = vis.type.schemas.metrics[0].name;
+    return aggs.bySchemaName(schemaName);
+  }, [vis.type.schemas.metrics[0].name, aggs, aggsLabels]);
```

Comment:
Why are we invalidating this memo on `aggsLabels`? It doesn't seem to be used. Same for other memos and effects down below

Reply:
Since `aggs` is defined in angular and it's mutable (any changes in `aggs` - adding/removing agg, aggType/aggField selection - don't substitute the link of `aggs`), react component doesn't respond. `aggsLabels` allows to identify when `aggs` is changed.
When the whole Vis Editor is reactified and `aggs` becomes immutable, we'll get rid of `aggsLabels` because `aggs` link will change correctly.